src/Components/Web/src/Virtualization/Virtualize.cs

```diff
@@ -0,0 +1,304 @@
+        /// </summary>
+        [Parameter]
+        public RenderFragment<PlaceholderContext>? Placeholder { get; set; }
+
+        /// <summary>
+        /// Gets the size of each item in pixels.
+        /// </summary>
+        [Parameter]
+        public float ItemSize { get; set; }
+
+        /// <summary>
+        /// Gets or sets the function providing items to the list.
+        /// </summary>
+        [Parameter]
+        public ItemsProviderDelegate<TItem> ItemsProvider { get; set; } = default!;
+
+        /// <summary>
+        /// Gets or sets the fixed item source.
+        /// </summary>
+        [Parameter]
+        public ICollection<TItem> Items { get; set; } = default!;
+
+        /// <inheritdoc />
+        protected override void OnParametersSet()
+        {
+            if (ItemSize <= 0)
+            {
+                throw new InvalidOperationException(
+                    $"{GetType()} requires a positive value for parameter '{nameof(ItemSize)}' to perform virtualization.");
+            }
+
+            if (ItemsProvider != null)
+            {
+                if (Items != null)
+                {
+                    throw new InvalidOperationException(
+                        $"{GetType()} can only accept one item source from its parameters. " +
+                        $"Do not supply both '{nameof(Items)}' and '{nameof(ItemsProvider)}'");
+                }
+
+                _itemsProvider = ItemsProvider;
+            }
+            else if (Items != null)
+            {
+                _itemsProvider = DefaultItemsProvider;
+            }
+            else
+            {
+                throw new InvalidOperationException(
+                    $"{GetType()} requires either the '{nameof(Items)}' or '{nameof(ItemsProvider)}' parameters to be specified " +
+                    $"and non-null.");
+            }
+
+            _itemTemplate = Item ?? ChildContent;
+            _placeholder = Placeholder ?? DefaultPlaceholder;
+        }
+
+        /// <inheritdoc />
+        protected override async Task OnAfterRenderAsync(bool firstRender)
+        {
+            if (firstRender)
+            {
+                _jsInterop = new VirtualizeJsInterop(this, JSRuntime);
+                await _jsInterop.InitAsync(_spacerBefore, _spacerAfter);
+            }
+        }
+
+        /// <inheritdoc />
+        protected override void BuildRenderTree(RenderTreeBuilder builder)
+        {
+            if (_refreshException != null)
+            {
+                throw _refreshException;
+            }
+
+            builder.OpenElement(0, "div");
+            builder.AddAttribute(1, "style", GetSpacerStyle(_itemsBefore));
+            builder.AddElementReferenceCapture(2, elementReference => _spacerBefore = elementReference);
+            builder.CloseElement();
+
+            // This is a rare case where it's valid for the sequence number to be programmatically incremented.
```

Comment:
I think it might be better to move this comment closer to line 154 (unless I"m missing somehting)